In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
visa_df= pd.read_csv(r"C:\Users\Kasipetaraviteja\Documents\Data_files\Visadataset.csv")
visa_df

,case_id,continent,education_of_employee,has_job_experience,requires_job_training,no_of_employees,yr_of_estab,region_of_employment,prevailing_wage,unit_of_wage,full_time_position,case_status
0,EZYV01,Asia,High School,N,N,14513,2007,West,592.2029,Hour,Y,Denied
1,EZYV02,Asia,Master's,Y,N,2412,2002,Northeast,83425.6500,Year,Y,Certified
2,EZYV03,Asia,Bachelor's,N,Y,44444,2008,West,122996.8600,Year,Y,Denied
3,EZYV04,Asia,Bachelor's,N,N,98,1897,West,83434.0300,Year,Y,Denied
4,EZYV05,Africa,Master's,Y,N,1082,2005,South,149907.3900,Year,Y,Certified
...,...,...,...,...,...,...,...,...,...,...,...,...
25475,EZYV25476,Asia,Bachelor's,Y,Y,2601,2008,South,77092.5700,Year,Y,Certified
25476,EZYV25477,Asia,High School,Y,N,3274,2006,Northeast,279174.7900,Year,Y,Certified
25477,EZYV25478,Asia,Master's,Y,N,1121,1910,South,146298.8500,Year,N,Certified
25478,EZYV25479,Asia,Master's,Y,Y,1918,1887,West,86154.7700,Year,Y,Certified


In [4]:
## seperate categorical and numerical columns

In [3]:
categorical = visa_df.select_dtypes(include='object').columns.tolist()
numerical = visa_df.select_dtypes(exclude='object').columns.tolist()
data= [categorical,numerical]
df  = pd.DataFrame(data,index=['cat','nume']).T
df

,cat,nume
0,case_id,no_of_employees
1,continent,yr_of_estab
2,education_of_employee,prevailing_wage
3,has_job_experience,None
4,requires_job_training,None
5,region_of_employment,None
6,unit_of_wage,None
7,full_time_position,None
8,case_status,None


## frequency table

In [8]:
visa_df['continent']

0          Asia
1          Asia
2          Asia
3          Asia
4        Africa
          ...  
25475      Asia
25476      Asia
25477      Asia
25478      Asia
25479      Asia
Name: continent, Length: 25480, dtype: object

In [11]:
visa_df   # Df
visa_df['continent'] 
cond = visa_df['continent'] == 'Asia'
np.sum(cond)

np.int64(16861)

### unique()

### nunique()

In [10]:
## for unique values
visa_df['continent'].unique()   # unique columns

array(['Asia', 'Africa', 'North America', 'Europe', 'South America',
       'Oceania'], dtype=object)

In [11]:
visa_df['continent'].nunique()   # no of unique values

6

- there are 25480 applicants from 6 continents

- number of applicants from each continent : **Frequency table**

In [12]:
# Number of aplicants from 'Asia

visa_df   # Df
visa_df['continent']   # one column
cond = visa_df['continent'] == 'Asia'   # column observation/ frequency   ( True, False)
# for calc no.of applicants from Asia
print(f'No.of applicants from Asia : {np.sum(cond)}')      # count
# for data
visa_df[cond]  # extraction it will data of asia applicants

No.of applicants from Asia : 16861


,case_id,continent,education_of_employee,has_job_experience,requires_job_training,no_of_employees,yr_of_estab,region_of_employment,prevailing_wage,unit_of_wage,full_time_position,case_status
0,EZYV01,Asia,High School,N,N,14513,2007,West,592.2029,Hour,Y,Denied
1,EZYV02,Asia,Master's,Y,N,2412,2002,Northeast,83425.6500,Year,Y,Certified
2,EZYV03,Asia,Bachelor's,N,Y,44444,2008,West,122996.8600,Year,Y,Denied
3,EZYV04,Asia,Bachelor's,N,N,98,1897,West,83434.0300,Year,Y,Denied
5,EZYV06,Asia,Master's,Y,N,2339,2012,South,78252.1400,Year,Y,Certified
...,...,...,...,...,...,...,...,...,...,...,...,...
25475,EZYV25476,Asia,Bachelor's,Y,Y,2601,2008,South,77092.5700,Year,Y,Certified
25476,EZYV25477,Asia,High School,Y,N,3274,2006,Northeast,279174.7900,Year,Y,Certified
25477,EZYV25478,Asia,Master's,Y,N,1121,1910,South,146298.8500,Year,N,Certified
25478,EZYV25479,Asia,Master's,Y,Y,1918,1887,West,86154.7700,Year,Y,Certified


In [13]:
# for all continents / all classes
con = visa_df['continent'] == 'Asia'  # one class
np.sum(con)

con = visa_df['continent'] == 'Africa'  # one class
np.sum(con)

con = visa_df['continent'] == 'North America'  # one class
np.sum(con)

con = visa_df['continent'] == 'Europe'  # one class
np.sum(con)

con = visa_df['continent'] == 'Asia'  # one class
np.sum(con)

np.int64(16861)

In [17]:
continents = visa_df['continent'].unique()
count = []
for i in continents:
    con = visa_df['continent'] == i  # one class
    count.append( np.sum(con) )
count

data = zip(continents,count)
df = pd.DataFrame(data)
df.columns = ['continent','freq']
df

,continent,freq
0,Asia,16861
1,Africa,551
2,North America,3292
3,Europe,3732
4,South America,852
5,Oceania,192


In [14]:
continents = visa_df['continent'].unique()
freq=[]
for i in continents:
    cond = visa_df['continent'] == i
    freq.append(np.sum(cond))
freq

[np.int64(16861),
 np.int64(551),
 np.int64(3292),
 np.int64(3732),
 np.int64(852),
 np.int64(192)]

In [15]:
# dataframe
pd.DataFrame(zip(continents,freq), columns =['class','frequency'])

,class,frequency
0,Asia,16861
1,Africa,551
2,North America,3292
3,Europe,3732
4,South America,852
5,Oceania,192


##shortcut

## Value_Counts()

- it counts == freq

In [39]:
visa_df['continent'].value_counts()

continent
Asia             16861
Europe            3732
North America     3292
South America      852
Africa             551
Oceania            192
Name: count, dtype: int64

In [ ]:
### creating a df from above series

In [43]:
keys = visa_df['continent'].value_counts().keys().tolist()
values = visa_df['continent'].value_counts().values.tolist()
values

[16861, 3732, 3292, 852, 551, 192]

In [45]:
pd.DataFrame(zip(keys,values),columns=['continent','frequency'])

,continent,frequency
0,Asia,16861
1,Europe,3732
2,North America,3292
3,South America,852
4,Africa,551
5,Oceania,192


## group by

In [7]:
visa_df.groupby(by='case_status').count()['requires_job_training']  

# it is first group  case status 
#           certified   :   ? count of rows
#           denied      :  ?  count of rows

# then for which cloumn you want the result   ['requires_job_training']

case_status
Certified    17018
Denied        8462
Name: requires_job_training, dtype: int64

In [5]:
visa_df

,case_id,continent,education_of_employee,has_job_experience,requires_job_training,no_of_employees,yr_of_estab,region_of_employment,prevailing_wage,unit_of_wage,full_time_position,case_status
0,EZYV01,Asia,High School,N,N,14513,2007,West,592.2029,Hour,Y,Denied
1,EZYV02,Asia,Master's,Y,N,2412,2002,Northeast,83425.6500,Year,Y,Certified
2,EZYV03,Asia,Bachelor's,N,Y,44444,2008,West,122996.8600,Year,Y,Denied
3,EZYV04,Asia,Bachelor's,N,N,98,1897,West,83434.0300,Year,Y,Denied
4,EZYV05,Africa,Master's,Y,N,1082,2005,South,149907.3900,Year,Y,Certified
...,...,...,...,...,...,...,...,...,...,...,...,...
25475,EZYV25476,Asia,Bachelor's,Y,Y,2601,2008,South,77092.5700,Year,Y,Certified
25476,EZYV25477,Asia,High School,Y,N,3274,2006,Northeast,279174.7900,Year,Y,Certified
25477,EZYV25478,Asia,Master's,Y,N,1121,1910,South,146298.8500,Year,N,Certified
25478,EZYV25479,Asia,Master's,Y,Y,1918,1887,West,86154.7700,Year,Y,Certified


In [58]:
visa_df.groupby(by='region_of_employment').count()['continent']

region_of_employment
Island        375
Midwest      4307
Northeast    7195
South        7017
West         6586
Name: continent, dtype: int64

In [61]:
visa_df['region_of_employment'].value_counts()

region_of_employment
Northeast    7195
South        7017
West         6586
Midwest      4307
Island        375
Name: count, dtype: int64

In [18]:
region_emp = visa_df['region_of_employment'].unique().tolist()
freq = []
for i in region_emp:
    cond = visa_df['region_of_employment'] == i
    freq.append(np.sum(cond))
freq

[np.int64(6586), np.int64(7195), np.int64(7017), np.int64(4307), np.int64(375)]

In [19]:
# data frame
df = pd.DataFrame(zip(region_emp,freq), columns=['regionOfEmployment','frequency'])
df

,regionOfEmployment,frequency
0,West,6586
1,Northeast,7195
2,South,7017
3,Midwest,4307
4,Island,375


In [66]:
cond = visa_df['region_of_employment'] == 'South'
np.sum(cond)

np.int64(7017)

In [31]:
data = visa_df['region_of_employment'].value_counts()
keys = data.keys()
values = data.values
df = pd.DataFrame(zip(keys,values), columns=['region_of_employment','freq'])
df

,region_of_employment,freq
0,Northeast,7195
1,South,7017
2,West,6586
3,Midwest,4307
4,Island,375


region_of_employment
Northeast    7195
South        7017
West         6586
Midwest      4307
Island        375
Name: count, dtype: int64